pkg/peer/subscription/video.go

```diff
@@ -0,0 +1,145 @@
+package subscription
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/matrix-org/waterfall/pkg/common"
+	"github.com/pion/rtcp"
+	"github.com/pion/rtp"
+	"github.com/pion/webrtc/v3"
+	"github.com/sirupsen/logrus"
+)
+
+type RequestKeyFrameFn = func(track common.TrackInfo, simulcast common.SimulcastLayer)
+
+type VideoSubscription struct {
+	rtpSender *webrtc.RTPSender
+	rtpTrack  *webrtc.TrackLocalStaticRTP
+
+	info           common.TrackInfo
+	currentLayer   common.SimulcastLayer
+	packetRewriter *PacketRewriter
+
+	controller        SubscriptionController
+	requestKeyFrameFn RequestKeyFrameFn
+	watchdog          *common.WatchdogChannel
+	logger            *logrus.Entry
+}
+
+func NewVideoSubscription(
+	info common.TrackInfo,
+	simulcast common.SimulcastLayer,
+	controller SubscriptionController,
+	requestKeyFrameFn RequestKeyFrameFn,
+	logger *logrus.Entry,
+) (*VideoSubscription, error) {
+	// Create a new track.
+	rtpTrack, err := webrtc.NewTrackLocalStaticRTP(info.Codec, info.TrackID, info.StreamID)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create track: %s", err)
+	}
+
+	rtpSender, err := controller.AddTrack(rtpTrack)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to add track: %s", err)
+	}
+
+	// This is the SSRC that all outgoing (rewritten) packets will have.
+	outgoingSSRC := uint32(rtpSender.GetParameters().Encodings[0].SSRC)
+
+	// Create a subscription.
+	subscription := &VideoSubscription{
+		rtpSender,
+		rtpTrack,
+		info,
+		simulcast,
+		NewPacketRewriter(outgoingSSRC),
+		controller,
+		requestKeyFrameFn,
+		nil,
+		logger,
+	}
+
+	// Configure watchdog for the subscription so that we know when we don't receive any new frames.
+	watchdogConfig := common.WatchdogConfig{
+		Timeout: 2 * time.Second,
+		OnTimeout: func() {
+			logger.Warnf("No RTP on subscription for %s (%s)", subscription.info.TrackID, subscription.currentLayer)
+			subscription.requestKeyFrame()
```

Comment:
Was this going to switch down to whatever layer is still getting packets?

Reply:
Yeah, right, I've specifically created this helper `requestKeyFrame()` which uses `subscription.info` and `subscription.simulcast`, i.e. making sure that we use the actual current layer and not the layer that we used when creating the initial subscription.

Reply:
ah, I missed that this was doing this